helpers/src/db/pagination.rs

```diff
@@ -0,0 +1,69 @@
+use diesel::pg::Pg;
+use diesel::prelude::*;
+use diesel::query_builder::*;
+use diesel::query_dsl::methods::LoadQuery;
+use diesel::sql_types::BigInt;
+use diesel::QueryId;
+
+const DEFAULT_PER_PAGE: i64 = 10;
+
+pub trait Paginate: Sized {
+    fn paginate(self, page: i64) -> Paginated<Self>;
+}
+
+impl<T> Paginate for T {
+    fn paginate(self, page: i64) -> Paginated<Self> {
+        Paginated {
+            query: self,
+            per_page: DEFAULT_PER_PAGE,
+            page,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, QueryId)]
+pub struct Paginated<T> {
+    query: T,
+    page: i64,
+    per_page: i64,
+}
+
+impl<T> Paginated<T> {
+    pub fn per_page(self, per_page: i64) -> Self {
+        Paginated { per_page, ..self }
+    }
+
+    pub fn load_and_count_pages<U>(self, conn: &PgConnection) -> QueryResult<(Vec<U>, i64)>
+    where
+        Self: LoadQuery<PgConnection, (U, i64)>,
+    {
+        let per_page = self.per_page;
+        let results = self.load::<(U, i64)>(conn)?;
+        let total = results.get(0).map(|x| x.1).unwrap_or(0);
+        let records = results.into_iter().map(|x| x.0).collect();
+        let total_pages = (total as f64 / per_page as f64).ceil() as i64;
+        Ok((records, total_pages))
+    }
+}
+
+impl<T: Query> Query for Paginated<T> {
+    type SqlType = (T::SqlType, BigInt);
+}
+
+impl<T> RunQueryDsl<PgConnection> for Paginated<T> {}
+
+impl<T> QueryFragment<Pg> for Paginated<T>
+where
+    T: QueryFragment<Pg>,
+{
+    fn walk_ast(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        out.push_sql("SELECT *, COUNT(*) OVER () FROM (");
+        self.query.walk_ast(out.reborrow())?;
+        out.push_sql(") t LIMIT ");
+        out.push_bind_param::<BigInt, _>(&self.per_page)?;
+        out.push_sql(" OFFSET ");
```

Comment:
Since you are using plain SQL `LIMIT` and `OFFSET` I wanted to mention that I planned to use the the methods for that:
* [`.limit(limit: i64)`](https://docs.rs/diesel/1.4.5/diesel/query_dsl/trait.QueryDsl.html#method.limit)
* [`.offset(offset: i64)`](https://docs.rs/diesel/1.4.5/diesel/query_dsl/trait.QueryDsl.html#method.offset)

Maybe you can also use that for the trait to make it easier understandable?

Reply:
Not really. These methods are implemented in their own query fragments. When you use this trait they are not required anymore. Internally they have the same logic than this trait, just that this one also does the counting and pagination for you.

Reply:
But can't we use the methods to simplify the code on our side?
And do sth like (pseudocode)? :
```rust
self.query.limit(&self.per_page);
self.query.offset((self.page - 1) * self.per_page);
```

Reply:
The `limit` and `offset` are query methods, which only work on queries. The `self.query` contains the query, which we created before the `.paginate`, but we have to execute it as subquery within the query we are building here. The `out` contains the syntax tree of the created query and has the type AstPass, which does not implement them. This part of the QueryFragment is where you have to write SQL. I don't see a way around it.